src/abi.md

```diff
@@ -1,21 +1,209 @@
+
+> [!NOTE]
+> [`core::marker::Sized`] types have a *metadata type* of `()`.
+
+> [!NOTE]
+> With transitivity, this applies regardless of the mutability of either pointer type
+
+```rust
+unsafe fn foo(x: *mut i32){
+   unsafe{x.write(5);}
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
+    let mut val = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.reference-box]
+The types [`&T`], [`&mut T`], [`alloc::boxed::Box<T>`], and [`core::ptr::NonNull<T>`], are *abi compatible* with `*const T`
+
+> [!NOTE]
+> With transitivity, they are also *abi compatible* with each other, and with `*mut T`, as well as references/`Box` to different types that have the same *metadata type*.
+
+```rust
+fn foo(x: &mut i32){
+   *x = 5;
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
+    let mut val = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.core]
+The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.
+
+r[abi.compatibility.transparent]
+A [`struct`] declared with the `transparent` representation is *abi compatible* with its field that does not have size 0 and alignment 1, if such a field exists.
+
+r[abi.compatibilty.zst]
+Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.
+
+r[abi.compatibility.option]
+If `T` is a type listed in [layout.enum.option](https://doc.rust-lang.org/stable/core/option/index.html#representation), then given `S` is a type with size 0 and alignment 1, `T` is *abi compatible* with the types [`core::option::Option<T>`], [`core::result::Result<T,S>`], and [`core::result::Result<S,T>`].
+
+
+r[abi.compatibility.fn-ptr]
+An [`fn`-ptr type] `T` is compatible with an [`fn`-ptr type] `U` if `T` and `U` have *abi compatible* tags.
+
+r[abi.compatibility.extern-tag]
+Two [abi tags][abi tag] are *abi compatible* if:
+* They are the same string, or
+* One tag is `"X"`, and the other is `"X-unwind"`
+
+r[abi.compatibility.signature]
+Two function signatures are compatible if:
+* The [abi tags][abi tag] of both signatures are *abi compatible*,
+* They have the same number of parameters, excluding C-varargs,
+* Each parameter of both signatures, in order, are *abi compatible*, and
+* Either both signatures have C-varargs, or neither signature does.
+
+> [!NOTE]
+> A signature is compatible with itself.
+
+r[abi.compatibility.simd-abi]
+A type has *simd abi requirements* if:
+* It is a type declared with the standard-library repr-attrbute `simd`,
+* It is a aggregate type[^1], which has a type with *simd abi requirements* as a field.
```

Comment:
Can you use labels instead of numbers for footnotes? It makes it a little easier to avoid collisions, and makes it easier to see semantically what a footnote is referring to.